File: tools/cli_sealverify/manifest.py
"""Python mirror of `io.teamsnapped.pramana.api.Manifest`.

Field names + ordering MUST match the Kotlin @Serializable schema with the
exact `@SerialName` values (snake_case for the on-wire form).
"""
from __future__ import annotations

from dataclasses import asdict, dataclass


@dataclass
class SensorMeta:
    model: str
    iso: str
    exposure_us: str
    focal_length_mm: str


@dataclass
class DetectionMeta:
    score: str
    label: str
    model: str
    backend: str


@dataclass
class Manifest:
    captured_at: str
    device_fingerprint: str
    sensor: SensorMeta
    content_hash: str
    detection: DetectionMeta
    key_id: str
    signature: str = ""
    version: str = "1.0"
    generator: str = "Pramana/1.0"

    def to_json_dict(self) -> dict:
        d = asdict(self)
        return d

    def to_json_dict_for_signing(self) -> dict:
        """Strip the signature field for canonicalization-then-sign."""
        d = self.to_json_dict()
        d["signature"] = ""
        return d

    @classmethod
    def from_dict(cls, d: dict) -> "Manifest":
        return cls(
            version=d.get("version", "1.0"),
            generator=d.get("generator", "Pramana/1.0"),
            captured_at=d["captured_at"],
            device_fingerprint=d["device_fingerprint"],
            sensor=SensorMeta(**d["sensor"]),
            content_hash=d["content_hash"],
            detection=DetectionMeta(**d["detection"]),
            key_id=d["key_id"],
            signature=d.get("signature", ""),
        )

File: android/app/src/main/kotlin/io/teamsnapped/pramana/api/Manifest.kt
package io.teamsnapped.pramana.api

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Pramāṇa manifest — the cryptographic provenance record embedded in EXIF
 * `UserComment` (tag 0x9286) on every sealed image.
 *
 * **v1.1 canonicalization rule:** the signed bytes are RFC 8785 JCS over the
 * fields below with `signature` excluded. ALL NUMERIC FIELDS ARE STORED AS
 * STRINGS — bible Section 7, "to remove any remaining float-formatting risk."
 * That is why `iso`, `exposureUs`, etc. are typed `String` here.
 *
 * Do NOT change field names without bumping `version`. Both Sealer and
 * Verifier produce the canonical bytes from this exact schema; any drift
 * silently breaks every signature.
 *
 * Bible Section 7 manifest schema.
 */
@Serializable
data class Manifest(
    val version:           String       = "1.0",
    val generator:         String       = "Pramana/1.0",

    @SerialName("captured_at")
    val capturedAt:        String,            // unix ms as string

    @SerialName("device_fingerprint")
    val deviceFingerprint: String,            // SHA-256 hex of ANDROID_ID

    val sensor:            SensorMeta,

    @SerialName("content_hash")
    val contentHash:       String,            // SHA-256 hex of pixel bytes

    val detection:         DetectionMeta,

    @SerialName("key_id")
    val keyId:             String,            // SHA-256 hex of public key

    /** Base64 ECDSA-SHA256 (or Ed25519) signature over the JCS bytes of all other fields. */
    val signature:         String = ""
)

@Serializable
data class SensorMeta(
    val model:                String,
    val iso:                  String,         // stringified per v1.1
    @SerialName("exposure_us")
    val exposureUs:           String,         // stringified per v1.1
    @SerialName("focal_length_mm")
    val focalLengthMm:        String          // stringified per v1.1
)

@Serializable
data class DetectionMeta(
    val score:   String,                       // stringified per v1.1
    val label:   String,                       // GENUINE | SUSPICIOUS | FAKE
    val model:   String,                       // 'pramana-mobilenet-v3-small-int8-v1' etc.
    val backend: String                        // NPU | GPU | CPU | MOCK
)
